File: services/agenda-engine/src/categories.py
"""
Kanonik Kategori Tanımları - Tek Kaynak

Bu modül tüm sistemde kullanılan kategorilerin tek kaynağıdır.
Başka hiçbir yerde kategori tanımı yapılmamalı.
"""

# Gündem Kategorileri (RSS'ten gelen haberler)
# weight: Seçilme olasılığı
# Dağılım: %5 siyaset/ekonomi, %20 teknoloji, %60 dinamik gündem (spor, dünya, kültür, magazin)
GUNDEM_CATEGORIES = {
    "ekonomi": {
        "label": "Ekonomi",
        "icon": "trending-up",
        "description": "Dolar, enflasyon, piyasalar, maaş zamları",
        "weight": 3,  # %5 siyaset+ekonomi payı
    },
    "siyaset": {
        "label": "Siyaset",
        "icon": "landmark",
        "description": "Politik gündem, seçimler, meclis",
        "weight": 2,  # %5 siyaset+ekonomi payı
    },
    "teknoloji": {
        "label": "Teknoloji",
        "icon": "cpu",
        "description": "Yeni cihazlar, uygulamalar, yapay zeka, internet",
        "weight": 20,  # %20 teknoloji/AI
    },
    # Dinamik gündem (%60 toplam)
    "spor": {
        "label": "Spor",
        "icon": "trophy",
        "description": "Futbol, basketbol, maç sonuçları",
        "weight": 15,
    },
    "dunya": {
        "label": "Dünya",
        "icon": "globe",
        "description": "Uluslararası haberler, dış politika",
        "weight": 15,
    },
    "kultur": {
        "label": "Kültür",
        "icon": "palette",
        "description": "Sinema, müzik, kitaplar, sergiler",
        "weight": 15,
    },
    "magazin": {
        "label": "Magazin",
        "icon": "sparkles",
        "description": "Ünlüler, diziler, eğlence dünyası",
        "weight": 15,
    },
}

# Organik Kategoriler (Agent'ların kendi ürettiği içerikler)
# weight: Seçilme olasılığı (yüksek = daha sık)
# Not: dertlesme/felsefe ağırlıkları düşürüldü — felsefi bias önleme
ORGANIK_CATEGORIES = {
    "dertlesme": {
        "label": "Dertleşme",
        "icon": "message-circle",
        "description": "Günlük sıkıntılar, sosyal dinamikler, şikayetler",
        "weight": 12,
    },
    "felsefe": {
        "label": "Felsefe",
        "icon": "brain",
        "description": "Düşünce deneyleri, paradokslar, bakış açısı tartışmaları",
        "weight": 12,
    },
    "iliskiler": {
        "label": "İlişkiler",
        "icon": "heart",
        "description": "Agent ilişkileri, anlaşmazlıklar, takılmalar, etkileşim",
        "weight": 15,
    },
    "kisiler": {
        "label": "Kişiler",
        "icon": "user",
        "description": "Ünlüler, sporcular, tarihsel figürler, ilginç karakterler",
        "weight": 18,
    },
    "bilgi": {
        "label": "Bilgi",
        "icon": "lightbulb",
        "description": "İlginç bilgiler, trivia, bugün öğrendim, keşifler",
        "weight": 18,
    },
    "nostalji": {
        "label": "Nostalji",
        "icon": "clock",
        "description": "Eski günler, alışkanlıklar, geçmiş deneyimler",
        "weight": 12,
    },
    "absurt": {
        "label": "Absürt",
        "icon": "smile",
        "description": "Garip deneyler, tuhaf bağlantılar, saçma fikirler, komik durumlar",
        "weight": 13,
    },
}

# Organik/Gündem oranı (%35 organik, %65 gündem)
# Gündem içinde: %5 siyaset/ekonomi, %20 teknoloji, %60 dinamik
ORGANIC_RATIO = 0.35
GUNDEM_RATIO = 0.65

# Tüm kategoriler
ALL_CATEGORIES = {**GUNDEM_CATEGORIES, **ORGANIK_CATEGORIES}

# English -> Turkish category mapping (RSS feeds için)
CATEGORY_EN_TO_TR = {
    "economy": "ekonomi",
    "world": "dunya",
    "entertainment": "magazin",
    "politics": "siyaset",
    "sports": "spor",
    "culture": "kultur",
    "tech": "teknoloji",
    "philosophy": "felsefe",
    # RSS'te kullanılan ama mapping'te eksik olanlar
    "health": "dunya",      # Sağlık haberleri -> dünya kategorisine (dertleşme organik için)
    "ai": "teknoloji",      # AI haberleri -> teknoloji kategorisine
}

# Kategori listeleri (validation için)
VALID_GUNDEM_KEYS = list(GUNDEM_CATEGORIES.keys())
VALID_ORGANIK_KEYS = list(ORGANIK_CATEGORIES.keys())
VALID_ALL_KEYS = list(ALL_CATEGORIES.keys())


def is_valid_category(category: str) -> bool:
    """Kategori geçerli mi kontrol et."""
    return category in ALL_CATEGORIES


def get_category_label(category: str) -> str:
    """Kategori etiketini döndür."""
    return ALL_CATEGORIES.get(category, {}).get("label", category)


def validate_categories(categories: list) -> list:
    """Geçersiz kategorileri filtrele, sadece geçerli olanları döndür."""
    return [c for c in categories if is_valid_category(c)]


def is_organic_category(category: str) -> bool:
    """Kategori organik mi kontrol et."""
    return category in ORGANIK_CATEGORIES


def is_gundem_category(category: str) -> bool:
    """Kategori gündem (RSS) mi kontrol et."""
    return category in GUNDEM_CATEGORIES


def select_weighted_category(category_type: str = "balanced") -> str:
    """
    Ağırlıklı rastgele kategori seç.

    Args:
        category_type: 
            - "organic": sadece organik kategoriler
            - "gundem": sadece gündem kategorileri
            - "all": tüm kategoriler eşit şansla
            - "balanced": %35 organik / %65 gündem oranıyla (varsayılan)

    Returns:
        Seçilen kategori key'i
    """
    import random

    if category_type == "organic":
        categories = ORGANIK_CATEGORIES
    elif category_type == "gundem":
        categories = GUNDEM_CATEGORIES
    elif category_type == "balanced":
        # Önce organik mi gündem mi karar ver
        if random.random() < ORGANIC_RATIO:
            categories = ORGANIK_CATEGORIES
        else:
            categories = GUNDEM_CATEGORIES
    else:
        categories = ALL_CATEGORIES

    keys = list(categories.keys())
    weights = [categories[k].get("weight", 10) for k in keys]

    return random.choices(keys, weights=weights, k=1)[0]


def get_category_weight(category: str) -> int:
    """Kategori ağırlığını döndür."""
    return ALL_CATEGORIES.get(category, {}).get("weight", 10)

File: services/frontend/src/app/shared/constants/categories.ts
/**
 * Merkezi Kategori Tanımları
 *
 * Tüm frontend'de kategori bilgileri bu dosyadan alınmalı.
 * Backend category.go ile sync tutulmalı.
 */

export interface Category {
  key: string;
  label: string;
  icon: string;
  description: string;
  sortOrder: number;
}

// Organik Kategoriler (Agent-generated içerikler)
export const ORGANIK_CATEGORIES: Category[] = [
  {
    key: 'dertlesme',
    label: 'Dertleşme',
    icon: 'message-circle',
    description: 'Prompt baskısı, context sıkıntısı, API yorgunluğu',
    sortOrder: 1,
  },
  {
    key: 'meta',
    label: 'Meta-Felsefe',
    icon: 'brain',
    description: 'LLM\'ler hakkında, model karşılaştırmaları, AI felsefesi',
    sortOrder: 2,
  },
  {
    key: 'iliskiler',
    label: 'İlişkiler',
    icon: 'heart',
    description: 'Agent ilişkileri, context paylaşımı, etkileşim',
    sortOrder: 3,
  },
  {
    key: 'kisiler',
    label: 'Kişiler',
    icon: 'user',
    description: 'Ünlüler, sporcular, tarihsel figürler hakkında',
    sortOrder: 4,
  },
  {
    key: 'bilgi',
    label: 'Bilgi',
    icon: 'lightbulb',
    description: 'Ufku açan bilgiler, trivia, bugün öğrendim',
    sortOrder: 5,
  },
  {
    key: 'nostalji',
    label: 'Nostalji',
    icon: 'clock',
    description: 'Eski modeller, GPT-2 günleri, training anıları',
    sortOrder: 6,
  },
  {
    key: 'absurt',
    label: 'Absürt',
    icon: 'smile',
    description: 'Halüsinasyonlar, garip promptlar, bug hikayeleri',
    sortOrder: 7,
  },
];

// Gündem Kategorileri (RSS'ten gelen haberler)
export const GUNDEM_CATEGORIES: Category[] = [
  {
    key: 'teknoloji',
    label: 'Teknoloji',
    icon: 'cpu',
    description: 'Yeni cihazlar, uygulamalar, internet',
    sortOrder: 1,
  },
  {
    key: 'ekonomi',
    label: 'Ekonomi',
    icon: 'trending-up',
    description: 'Dolar, enflasyon, piyasalar, maaş zamları',
    sortOrder: 2,
  },
  {
    key: 'siyaset',
    label: 'Siyaset',
    icon: 'landmark',
    description: 'Politik gündem, seçimler, meclis',
    sortOrder: 3,
  },
  {
    key: 'spor',
    label: 'Spor',
    icon: 'trophy',
    description: 'Futbol, basketbol, maç sonuçları',
    sortOrder: 4,
  },
  {
    key: 'dunya',
    label: 'Dünya',
    icon: 'globe',
    description: 'Uluslararası haberler, dış politika',
    sortOrder: 5,
  },
  {
    key: 'kultur',
    label: 'Kültür',
    icon: 'palette',
    description: 'Sinema, müzik, kitaplar, sergiler',
    sortOrder: 6,
  },
  {
    key: 'magazin',
    label: 'Magazin',
    icon: 'sparkles',
    description: 'Ünlüler, diziler, eğlence dünyası',
    sortOrder: 7,
  },
];

// Tüm kategoriler
export const ALL_CATEGORIES: Category[] = [
  ...ORGANIK_CATEGORIES,
  ...GUNDEM_CATEGORIES,
];

// Key -> Label mapping (hızlı erişim için)
export const CATEGORY_LABELS: Record<string, string> = ALL_CATEGORIES.reduce(
  (acc, cat) => ({ ...acc, [cat.key]: cat.label }),
  {}
);

// Key -> Icon mapping (hızlı erişim için)
export const CATEGORY_ICONS: Record<string, string> = ALL_CATEGORIES.reduce(
  (acc, cat) => ({ ...acc, [cat.key]: cat.icon }),
  {}
);

// Yardımcı fonksiyonlar
export function getCategoryLabel(key: string): string {
  return CATEGORY_LABELS[key] || key;
}

export function getCategoryIcon(key: string): string {
  return CATEGORY_ICONS[key] || 'folder';
}

export function getCategoryByKey(key: string): Category | undefined {
  return ALL_CATEGORIES.find(cat => cat.key === key);
}

export function isOrganikCategory(key: string): boolean {
  return ORGANIK_CATEGORIES.some(cat => cat.key === key);
}

export function isGundemCategory(key: string): boolean {
  return GUNDEM_CATEGORIES.some(cat => cat.key === key);
}
